File: platforms/android/app/src/main/java/com/streamvue/player/data/PlexAccount.kt
package com.streamvue.player.data

import com.google.gson.Gson
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import java.net.URI
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.Locale

data class PlexPinChallenge(
    val id: Int,
    val code: String,
    val authorizationUrl: String,
    val expiresAt: Instant
)

data class PlexServerConnectionChoice(
    val url: String,
    val isLocal: Boolean,
    val isRelay: Boolean,
    val isSecure: Boolean,
    val isIpv6: Boolean
)

data class PlexDiscoveredServer(
    val serverId: String,
    val name: String,
    val isOwned: Boolean,
    val connections: List<PlexServerConnectionChoice>
) {
    val preferredConnection: PlexServerConnectionChoice?
        get() = connections.firstOrNull()
}

/** An opaque, short-lived discovery lease. It never contains a Plex token. */
data class PlexServerDiscovery(
    val sessionId: String,
    val servers: List<PlexDiscoveredServer>,
    val expiresAt: Instant
)

internal data class PlexAccountToken(
    val value: String,
    val expiresAt: Instant?
)

internal data class PlexAccountServerSecret(
    val server: PlexDiscoveredServer,
    val accessToken: String
)

internal data class PlexAccountDiscoverySecret(
    val servers: List<PlexAccountServerSecret>,
    val expiresAt: Instant
)

internal interface PlexDeviceSigner {
    val publicJwk: JsonObject
    fun sign(claims: Map<String, Any>): String
}

/** Implements Plex's signed strong-PIN account flow without retaining account tokens. */
internal class PlexAccountClient(
    private val transport: MediaCenterTransport,
    private val signer: PlexDeviceSigner,
    private val device: MediaCenterDeviceIdentity,
    private val gson: Gson = Gson(),
    private val now: () -> Instant = Instant::now
) {
    fun createPin(): PlexPinChallenge {
        val publicKey = validatePublicJwk(signer.publicJwk)
        val body = JsonObject().apply {
            add("jwk", publicKey)
            addProperty("strong", true)
        }
        val payload = json(
            MediaHttpRequest(
                method = MediaHttpMethod.POST,
                url = URI("$CLIENTS_BASE_URL/pins"),
                headers = headers + ("Content-Type" to "application/json"),
                body = gson.toJson(body).toByteArray(Charsets.UTF_8)
            )
        ).objectValue()
        val id = payload.integer("id")?.takeIf { it > 0 }
            ?: error("Plex returned an incomplete sign-in challenge.")
        val code = payload.text("code")?.let {
            MediaCenterUrlPolicy.requireIdentifier(it, "Plex sign-in code")
        } ?: error("Plex returned an incomplete sign-in challenge.")
        val createdAt = now()
        val expiresAt = expiry(payload, createdAt) ?: createdAt.plus(5, ChronoUnit.MINUTES)
        require(expiresAt.isAfter(createdAt)) { "Plex returned an expired sign-in challenge." }
        return PlexPinChallenge(
            id = id,
            code = code,
            authorizationUrl = authorizationUrl(code),
            expiresAt = expiresAt
        )
    }

    fun claimPin(challenge: PlexPinChallenge): PlexAccountToken? {
        require(challenge.id > 0 && challenge.expiresAt.isAfter(now())) {
            "The Plex sign-in request expired. Start a new sign-in."
        }
        val issuedAt = now().epochSecond
        require(issuedAt > 0) { "The device clock is invalid." }
        val proof = validateCompactJwt(
            signer.sign(
                sortedMapOf(
                    "aud" to "plex.tv",
                    "exp" to issuedAt + 300,
                    "iat" to issuedAt,
                    "iss" to clientIdentifier
                )
            )
        )
        val payload = json(
            MediaHttpRequest(
                method = MediaHttpMethod.GET,
                url = MediaCenterUrlPolicy.appendQuery(
                    URI("$CLIENTS_BASE_URL/pins/${challenge.id}"),
                    mapOf("deviceJWT" to proof)
                ),
                headers = headers
            )
        ).objectValue()
        return accountToken(payload)
    }

    fun verifyAccountToken(rawToken: String) {
        val token = MediaCenterUrlPolicy.credential(rawToken)
        json(
            MediaHttpRequest(
                method = MediaHttpMethod.GET,
                url = URI("$ACCOUNT_BASE_URL/user"),
                headers = authenticatedHeaders(token)
            )
        )
    }

    fun discoverServers(rawToken: String): List<PlexAccountServerSecret> {
        val accountToken = MediaCenterUrlPolicy.credential(rawToken)
        val payload = json(
            MediaHttpRequest(
                method = MediaHttpMethod.GET,
                url = MediaCenterUrlPolicy.appendQuery(
                    URI("$CLIENTS_BASE_URL/resources"),
                    mapOf("includeHttps" to "1", "includeRelay" to "1", "includeIPv6" to "1")
                ),
                headers = authenticatedHeaders(accountToken)
            )
        )
        return payload.arrayValue().mapNotNull { parseServer(it, accountToken) }
    }

    private val clientIdentifier: String
        get() = MediaCenterUrlPolicy.requireIdentifier(device.deviceId, "Plex client")

    private val headers: Map<String, String>
        get() = mapOf(
            "Accept" to "application/json",
            "X-Plex-Client-Identifier" to clientIdentifier,
            "X-Plex-Product" to safeHeader(device.client, "StreamVue"),
            "X-Plex-Version" to safeHeader(device.version, "5.1.0")
        )

    private fun authenticatedHeaders(token: String): Map<String, String> =
        headers + ("X-Plex-Token" to MediaCenterUrlPolicy.credential(token))

    private fun json(request: MediaHttpRequest): JsonElement {
        require(
            request.url.scheme.equals("https", true) &&
                request.url.host?.lowercase(Locale.ROOT) in PROVIDER_HOSTS &&
                request.url.userInfo == null
        ) { "The Plex account request address is unsafe." }
        val response = transport.execute(request)
        require(response.status in 200..299) { "Plex sign-in returned HTTP ${response.status}." }
        require(response.body.isNotEmpty() && response.body.size <= MAX_RESPONSE_BYTES) {
            "Plex returned an invalid sign-in response."
        }
        return runCatching { JsonParser.parseString(response.body.toString(Charsets.UTF_8)) }
            .getOrElse { error("Plex returned an invalid sign-in response.") }
    }

    private fun parseServer(value: JsonElement, accountToken: String): PlexAccountServerSecret? {
        val raw = value.objectValue()
        val provides = raw.text("provides").orEmpty().split(',').mapTo(HashSet()) {
            it.trim().lowercase(Locale.ROOT)
        }
        if ("server" !in provides) return null
        val serverId = raw.text("clientIdentifier")?.let {
            runCatching { MediaCenterUrlPolicy.requireIdentifier(it, "Plex server") }.getOrNull()
        } ?: return null
        val accessToken = raw.text("accessToken")?.let {
            runCatching { MediaCenterUrlPolicy.credential(it) }.getOrNull()
        } ?: return null
        if (serverId.contains(accessToken) || serverId.contains(accountToken)) return null
        val redactedName = MediaCenterUrlPolicy.safeMetadata(raw.text("name"), accessToken, 256)
        val name = MediaCenterUrlPolicy.safeMetadata(redactedName, accountToken, 256) ?: return null
        val connections = raw.arrayAt("connections")
            .mapNotNull { parseConnection(it, listOf(accountToken, accessToken)) }
            .sortedBy(::connectionPriority)
        if (connections.isEmpty()) return null
        return PlexAccountServerSecret(
            server = PlexDiscoveredServer(
                serverId = serverId,
                name = name,
                isOwned = raw.boolean("owned"),
                connections = connections
            ),
            accessToken = accessToken
        )
    }

    private fun parseConnection(
        value: JsonElement,
        credentials: List<String>
    ): PlexServerConnectionChoice? {
        val raw = value.objectValue()
        var candidate = raw.text("uri")
        if (candidate == null) {
            val scheme = raw.text("protocol")?.lowercase(Locale.ROOT)
            val address = raw.text("address")
            val port = raw.integer("port")
            if (scheme in setOf("http", "https") && address != null && port in 1..65_535) {
                val host = if (':' in address && !address.startsWith('[')) "[$address]" else address
                candidate = "$scheme://$host:$port"
            }
        }
        val url = candidate?.let {
            runCatching { MediaCenterUrlPolicy.normalizeBaseUrl(it) }.getOrNull()
        } ?: return null
        if (credentials.any { it.isEmpty() || url.toASCIIString().contains(it) }) return null
        return PlexServerConnectionChoice(
            url = url.toASCIIString(),
            isLocal = raw.boolean("local"),
            isRelay = raw.boolean("relay"),
            isSecure = url.scheme.equals("https", true),
            isIpv6 = raw.boolean("IPv6") || ':' in url.host.orEmpty()
        )
    }

    private fun accountToken(payload: JsonObject): PlexAccountToken? {
        val token = (payload.text("authToken") ?: payload.text("auth_token"))?.let {
            runCatching { MediaCenterUrlPolicy.credential(it) }.getOrNull()
        } ?: return null
        return PlexAccountToken(token, expiry(payload, now()))
    }

    private fun expiry(payload: JsonObject, relativeTo: Instant): Instant? {
        val explicit = payload.text("expiresAt") ?: payload.text("expires_at")
        if (explicit != null) runCatching { Instant.parse(explicit) }.getOrNull()?.let { return it }
        val seconds = payload.number("expiresIn") ?: payload.number("expires_in") ?: return null
        if (seconds <= 0) return null
        return runCatching { relativeTo.plusSeconds(seconds) }.getOrNull()
    }

    private fun authorizationUrl(code: String): String {
        val query = linkedMapOf(
            "clientID" to clientIdentifier,
            "code" to code,
            "context[device][product]" to safeHeader(device.client, "StreamVue")
        ).entries.joinToString("&") { (name, value) ->
            "${encode(name)}=${encode(value)}"
        }
        return "https://app.plex.tv/auth#?$query"
    }

    private fun validatePublicJwk(untrusted: JsonObject): JsonObject {
        require(!untrusted.has("d")) { "A Plex public device key cannot contain private key material." }
        require(
            untrusted.text("kty") == "OKP" && untrusted.text("crv") == "Ed25519" &&
                untrusted.text("alg") == "EdDSA"
        ) { "Plex requires an Ed25519 device signing key." }
        val x = untrusted.text("x").orEmpty()
        require(x.matches(Regex("^[A-Za-z0-9_-]{40,64}$"))) {
            "The Plex device public key is invalid."
        }
        val kid = MediaCenterUrlPolicy.requireIdentifier(
            untrusted.text("kid").orEmpty(),
            "Plex device key"
        )
        return JsonObject().apply {
            addProperty("kty", "OKP")
            addProperty("crv", "Ed25519")
            addProperty("x", x)
            addProperty("kid", kid)
            addProperty("alg", "EdDSA")
        }
    }

    private fun validateCompactJwt(rawValue: String): String {
        val value = rawValue.trim()
        require(value.length <= 16_384 && value.matches(COMPACT_JWT)) {
            "The Plex device proof is invalid."
        }
        return value
    }

    private fun connectionPriority(value: PlexServerConnectionChoice): Int =
        (if (value.isSecure) 0 else 1_000) +
            (if (value.isLocal) 0 else 100) +
            (if (value.isRelay) 50 else 0) +
            (if (value.isIpv6) 1 else 0)

    private fun safeHeader(value: String, fallback: String): String = value
        .filterNot(Char::isISOControl)
        .trim()
        .take(256)
        .ifEmpty { fallback }

    private fun encode(value: String): String = URLEncoder
        .encode(value, StandardCharsets.UTF_8.name())
        .replace("+", "%20")

    private companion object {
        const val CLIENTS_BASE_URL = "https://clients.plex.tv/api/v2"
        const val ACCOUNT_BASE_URL = "https://plex.tv/api/v2"
        const val MAX_RESPONSE_BYTES = 2 * 1_024 * 1_024
        val PROVIDER_HOSTS = setOf("clients.plex.tv", "plex.tv")
        val COMPACT_JWT = Regex("^[A-Za-z0-9_-]+\\.[A-Za-z0-9_-]+\\.[A-Za-z0-9_-]+$")
    }
}

File: platforms/apple/Sources/OrbitalVueCore/PlexAccount.swift
import CryptoKit
import Foundation

/// A short-lived Plex sign-in challenge. It contains no account credential.
public struct PlexPinChallenge: Identifiable, Equatable, Sendable {
    public let id: Int
    public let code: String
    public let authorizationURL: URL
    public let expiresAt: Date

    init(id: Int, code: String, authorizationURL: URL, expiresAt: Date) {
        self.id = id
        self.code = code
        self.authorizationURL = authorizationURL
        self.expiresAt = expiresAt
    }
}

public struct PlexServerConnectionChoice: Identifiable, Equatable, Sendable {
    public let url: URL
    public let isLocal: Bool
    public let isRelay: Bool
    public let isSecure: Bool
    public let isIPv6: Bool

    public var id: String { url.absoluteString }

    init(url: URL, isLocal: Bool, isRelay: Bool, isSecure: Bool, isIPv6: Bool) {
        self.url = url
        self.isLocal = isLocal
        self.isRelay = isRelay
        self.isSecure = isSecure
        self.isIPv6 = isIPv6
    }
}

/// A sanitized account resource. The server-scoped token stays inside
/// `MediaCenterService` and is never exposed to SwiftUI or catalog storage.
public struct PlexDiscoveredServer: Identifiable, Equatable, Sendable {
    public let serverID: String
    public let name: String
    public let isOwned: Bool
    public let connections: [PlexServerConnectionChoice]

    public var id: String { serverID }
    public var preferredConnection: PlexServerConnectionChoice? { connections.first }

    init(
        serverID: String,
        name: String,
        isOwned: Bool,
        connections: [PlexServerConnectionChoice]
    ) {
        self.serverID = serverID
        self.name = name
        self.isOwned = isOwned
        self.connections = connections
    }
}

/// An in-memory discovery lease. Its opaque ID refers to credentials held only
/// by the service actor and expires even if the UI forgets to cancel it.
public struct PlexServerDiscovery: Equatable, Sendable {
    public let sessionID: String
    public let servers: [PlexDiscoveredServer]
    public let expiresAt: Date

    init(sessionID: String, servers: [PlexDiscoveredServer], expiresAt: Date) {
        self.sessionID = sessionID
        self.servers = servers
        self.expiresAt = expiresAt
    }
}

struct PlexAccountServerSecret: Sendable {
    let server: PlexDiscoveredServer
    let accessToken: String
}

struct PlexAccountDiscoverySecret: Sendable {
    let servers: [PlexAccountServerSecret]
    let expiresAt: Date
}

struct PlexAccountToken: Sendable {
    let value: String
    let expiresAt: Date?
}

struct PlexDeviceSigner: Sendable {
    private let privateKey: Curve25519.Signing.PrivateKey

    init() {
        privateKey = Curve25519.Signing.PrivateKey()
    }

    init(rawRepresentation: Data) throws {
        privateKey = try Curve25519.Signing.PrivateKey(rawRepresentation: rawRepresentation)
    }

    var rawRepresentation: Data { privateKey.rawRepresentation }

    var publicJWK: [String: String] {
        let publicKey = privateKey.publicKey.rawRepresentation
        let keyID = Data(SHA256.hash(data: publicKey)).base64URLEncodedString()
        return [
            "kty": "OKP",
            "crv": "Ed25519",
            "x": publicKey.base64URLEncodedString(),
            "kid": keyID,
            "alg": "EdDSA"
        ]
    }

    func sign(claims: [String: Any]) throws -> String {
        let keyID = publicJWK["kid"] ?? ""
        let header: [String: Any] = [
            "alg": "EdDSA",
            "kid": keyID,
            "typ": "JWT"
        ]
        let encodedHeader = try Self.encodedJSON(header)
        let encodedClaims = try Self.encodedJSON(claims)
        let signingInput = "\(encodedHeader).\(encodedClaims)"
        let signature = try privateKey.signature(for: Data(signingInput.utf8))
        return "\(signingInput).\(signature.base64URLEncodedString())"
    }

    private static func encodedJSON(_ value: [String: Any]) throws -> String {
        guard JSONSerialization.isValidJSONObject(value) else {
            throw MediaCenterError.invalidResponse
        }
        let data = try JSONSerialization.data(withJSONObject: value, options: [.sortedKeys])
        return data.base64URLEncodedString()
    }
}

/// Implements Plex's signed, strong-PIN account flow. Account tokens are used
/// only long enough to verify the account and fetch server-scoped resources.
struct PlexAccountClient: Sendable {
    private static let clientsBaseURL = URL(string: "https://clients.plex.tv/api/v2")!
    private static let accountBaseURL = URL(string: "https://plex.tv/api/v2")!
    private static let maximumResponseBytes = 2 * 1_024 * 1_024

    private let httpClient: any MediaCenterHTTPClient
    private let clientIdentifier: String
    private let product: String
    private let version: String
    private let now: @Sendable () -> Date

    init(
        httpClient: any MediaCenterHTTPClient,
        clientIdentifier: String,
        product: String = "OrbitalVue",
        version: String = "5.8.0",
        now: @escaping @Sendable () -> Date = Date.init
    ) throws {
        self.httpClient = httpClient
        self.clientIdentifier = try MediaCenterURLPolicy.requireIdentifier(
            clientIdentifier,
            label: "Plex client"
        )
        self.product = MediaCenterHeaderPolicy.applicationValue(product, fallback: "OrbitalVue")
        self.version = MediaCenterHeaderPolicy.applicationValue(version, fallback: "5.8.0")
        self.now = now
    }

    func createPin(signer: PlexDeviceSigner) async throws -> PlexPinChallenge {
        let body = try JSONSerialization.data(
            withJSONObject: ["jwk": signer.publicJWK, "strong": true],
            options: [.sortedKeys]
        )
        let payload = try await json(
            method: .post,
            url: endpoint("pins"),
            headers: headers.merging(["Content-Type": "application/json"]) { _, new in new },
            body: body
        )
        guard let id = payload.integer("id"), id > 0,
              let rawCode = payload.text("code") else {
            throw MediaCenterError.invalidResponse
        }
        let code = try MediaCenterURLPolicy.requireIdentifier(rawCode, label: "Plex sign-in code")
        let createdAt = now()
        let expiresAt = expiry(in: payload, relativeTo: createdAt)
            ?? createdAt.addingTimeInterval(5 * 60)
        guard expiresAt > createdAt,
              let authorizationURL = authorizationURL(code: code) else {
            throw MediaCenterError.invalidResponse
        }
        return PlexPinChallenge(
            id: id,
            code: code,
            authorizationURL: authorizationURL,
            expiresAt: expiresAt
        )
    }

    func claimPin(
        _ challenge: PlexPinChallenge,
        signer: PlexDeviceSigner
    ) async throws -> PlexAccountToken? {
        guard challenge.id > 0, challenge.expiresAt > now() else {
            throw MediaCenterError.accountSignInExpired
        }
        let issuedAt = Int(now().timeIntervalSince1970.rounded(.down))
        guard issuedAt > 0 else { throw MediaCenterError.invalidResponse }
        let proof = try signer.sign(claims: [
            "aud": "plex.tv",
            "iss": clientIdentifier,
            "iat": issuedAt,
            "exp": issuedAt + 300
        ])
        guard var components = URLComponents(
            url: endpoint("pins/\(challenge.id)"),
            resolvingAgainstBaseURL: false
        ) else { throw MediaCenterError.invalidResponse }
        components.queryItems = [URLQueryItem(name: "deviceJWT", value: proof)]
        guard let url = components.url else { throw MediaCenterError.invalidResponse }
        let payload = try await json(method: .get, url: url, headers: headers)
        return accountToken(in: payload)
    }

    func verifyAccountToken(_ rawToken: String) async throws {
        let token = try MediaCenterHeaderPolicy.credential(rawToken)
        _ = try await json(
            method: .get,
            url: accountEndpoint("user"),
            headers: authenticatedHeaders(token: token)
        )
    }

    func discoverServers(accountToken rawToken: String) async throws -> [PlexAccountServerSecret] {
        let token = try MediaCenterHeaderPolicy.credential(rawToken)
        guard var components = URLComponents(
            url: endpoint("resources"),
            resolvingAgainstBaseURL: false
        ) else { throw MediaCenterError.invalidResponse }
        components.queryItems = [
            URLQueryItem(name: "includeHttps", value: "1"),
            URLQueryItem(name: "includeRelay", value: "1"),
            URLQueryItem(name: "includeIPv6", value: "1")
        ]
        guard let url = components.url else { throw MediaCenterError.invalidResponse }
        let payload = try await json(
            method: .get,
            url: url,
            headers: authenticatedHeaders(token: token)
        )
        return payload.arrayValue.compactMap { parseServer($0, excluding: token) }
    }

    private var headers: [String: String] {
        [
            "Accept": "application/json",
            "X-Plex-Client-Identifier": clientIdentifier,
            "X-Plex-Product": product,
            "X-Plex-Version": version
        ]
    }

    private func authenticatedHeaders(token: String) -> [String: String] {
        headers.merging(["X-Plex-Token": token]) { _, new in new }
    }

    private func endpoint(_ path: String) -> URL {
        Self.clientsBaseURL.appendingPathComponent(path)
    }

    private func accountEndpoint(_ path: String) -> URL {
        Self.accountBaseURL.appendingPathComponent(path)
    }

    private func json(
        method: MediaCenterHTTPMethod,
        url: URL,
        headers: [String: String],
        body: Data? = nil
    ) async throws -> MediaCenterJSON {
        guard url.scheme == "https",
              ["clients.plex.tv", "plex.tv"].contains(url.host?.lowercased() ?? ""),
              url.user == nil, url.password == nil else {
            throw MediaCenterError.unsafeProviderURL
        }
        return try await MediaCenterAPI.json(
            MediaCenterHTTPRequest(
                method: method,
                url: url,
                headers: headers,
                body: body,
                maximumResponseBytes: Self.maximumResponseBytes
            ),
            using: httpClient
        )
    }

    private func authorizationURL(code: String) -> URL? {
        var query = URLComponents()
        query.queryItems = [
            URLQueryItem(name: "clientID", value: clientIdentifier),
            URLQueryItem(name: "code", value: code),
            URLQueryItem(name: "context[device][product]", value: product)
        ]
        guard let encodedQuery = query.percentEncodedQuery else { return nil }
        var components = URLComponents()
        components.scheme = "https"
        components.host = "app.plex.tv"
        components.path = "/auth"
        components.percentEncodedFragment = "?\(encodedQuery)"
        return components.url
    }

    private func accountToken(in payload: MediaCenterJSON) -> PlexAccountToken? {
        let object = payload.objectValue
        guard let rawToken = object.text("authToken") ?? object.text("auth_token"),
              let token = try? MediaCenterHeaderPolicy.credential(rawToken) else {
            return nil
        }
        return PlexAccountToken(value: token, expiresAt: expiry(in: payload, relativeTo: now()))
    }

    private func expiry(in payload: MediaCenterJSON, relativeTo date: Date) -> Date? {
        let object = payload.objectValue
        if let value = object.text("expiresAt") ?? object.text("expires_at"),
           let parsed = ISO8601DateFormatter().date(from: value) {
            return parsed
        }
        guard let seconds = object.integer("expiresIn") ?? object.integer("expires_in"),
              seconds > 0 else { return nil }
        return date.addingTimeInterval(TimeInterval(seconds))
    }

    private func parseServer(
        _ value: MediaCenterJSON,
        excluding accountToken: String
    ) -> PlexAccountServerSecret? {
        let object = value.objectValue
        let provides = Set(
            (object.text("provides") ?? "")
                .split(separator: ",")
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }
        )
        guard provides.contains("server"),
              let rawServerID = object.text("clientIdentifier"),
              let rawName = object.text("name"),
              let rawToken = object.text("accessToken"),
              let accessToken = try? MediaCenterHeaderPolicy.credential(rawToken),
              let serverID = try? MediaCenterURLPolicy.requireIdentifier(
                  rawServerID,
                  label: "Plex server"
              ),
              !serverID.contains(accessToken),
              !serverID.contains(accountToken) else { return nil }

        let serverRedactedName = MediaCenterTextPolicy.metadata(
            rawName,
            redacting: accessToken,
            maximumLength: 256
        )
        let name = MediaCenterTextPolicy.metadata(
            serverRedactedName,
            redacting: accountToken,
            maximumLength: 256
        )
        guard let safeName = name.nonEmptyMediaCenterText else { return nil }
        let connections = object.array("connections")
            .compactMap { parseConnection($0, excluding: [accountToken, accessToken]) }
            .sorted { connectionPriority($0) < connectionPriority($1) }
        guard !connections.isEmpty else { return nil }
        let server = PlexDiscoveredServer(
            serverID: serverID,
            name: safeName,
            isOwned: object.boolean("owned"),
            connections: connections
        )
        return PlexAccountServerSecret(server: server, accessToken: accessToken)
    }

    private func parseConnection(
        _ value: MediaCenterJSON,
        excluding credentials: [String]
    ) -> PlexServerConnectionChoice? {
        let object = value.objectValue
        var candidate = object.text("uri")
        if candidate == nil,
           let scheme = object.text("protocol")?.lowercased(),
           ["http", "https"].contains(scheme),
           let address = object.text("address"),
           let port = object.integer("port"),
           (1...65_535).contains(port) {
            let host = address.contains(":") && !address.hasPrefix("[") ? "[\(address)]" : address
            candidate = "\(scheme)://\(host):\(port)"
        }
        guard let candidate,
              let url = try? MediaCenterURLPolicy.normalizeBaseURL(candidate),
              credentials.allSatisfy({ !$0.isEmpty && !url.absoluteString.contains($0) }) else {
            return nil
        }
        return PlexServerConnectionChoice(
            url: url,
            isLocal: object.boolean("local"),
            isRelay: object.boolean("relay"),
            isSecure: url.scheme?.lowercased() == "https",
            isIPv6: object.boolean("IPv6") || (url.host?.contains(":") == true)
        )
    }

    private func connectionPriority(_ value: PlexServerConnectionChoice) -> Int {
        (value.isSecure ? 0 : 1_000)
            + (value.isLocal ? 0 : 100)
            + (value.isRelay ? 50 : 0)
            + (value.isIPv6 ? 1 : 0)
    }
}

private extension Data {
    func base64URLEncodedString() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}
